src/engines/CP2K_engine.py

```diff
@@ -0,0 +1,111 @@
+"""
+Engine implementation of CP2K
+"""
+from typing import Sequence
+import numpy as np
+from cp2k_input_tools.parser import CP2KInputParser
+
+from engines.abstract_engine import AbstractEngine, ShootingResult
+
+
+class CP2KEngine(AbstractEngine):
+    def __init__(self, inputs):
+        super().__init__(inputs)
+
+        self._atoms = None
+        with open(inputs["cp2k_inputs"]) as f:
+            parser = CP2KInputParser()
+            self.cp2k_inputs = parser.parse(f)
+
+    @property
+    def atoms(self) -> Sequence[str]:
+        if self._atoms is None:
+            # TODO: How does this handle coordinates linked in a separate file?
+            # Return the first two places for each coordinate entry
+            self._atoms = [entry[0:2] for entry in self._get_coord()]
+
+        return self._atoms
+
+    def set_positions(self, positions: np.ndarray) -> None:
+        # Check positions are valid by passing to base class
+        super().set_positions(positions)
+
+        # coords stored as list of "El x y z" strings, same as CP2K .inp file
+        coords = self._get_coord()
```

Comment:
Where are these `coords` being stored outside of this method?

Reply:
Everything is being stored in the data structure `self.cp2k_inputs`, which is what the cp2k input parser creates. This is true for both coordinates and velocities. That function is just a helper to get to that specific section of the data structure.

Something i thought about was keeping these values in a separate array and only translating them into the `cp2k_inputs` when necessary, but we're almost always immediately write out a new inputs file after assigning new positions/velocities so there isn't much to gain by doing so.